book/src/proposals/durable-tx-nonces.md

```diff
@@ -23,30 +23,59 @@ implemented by some CPU ISAs.
 
 ### Contract Mechanics
 
-TODO: svgbob this into a flowchart, add prose
+TODO: svgbob this into a flowchart
 
 ```text
 Start
 Create Account
   state = Uninitialized
-InitializeInstruction(hash)
-  if hash is recent
-    stored_hash = hash
+InitializeInstruction()
+  if !sysvar.recent_blockhashes.is_empty()
+    stored_hash = sysvar.recent_blockhashes[0]
     state = Initialized
     success
   else
     error
-NonceInstruction(spend_hash, next_hash)
+NonceInstruction(spend_hash) /* spend_hash is a stand in for a TBD sysvar
```

Comment:
why does Nonce() take a value?

Reply:
It's just a stand in 'til I figure out exposing the TX message's `recent_blockhash` field via sysvar

Reply:
I'm missing something, I think.  why is it important for the Nonce program to verify that the TX is using its current nonce value?  I thought that check would be the runtime's job.

Reply:
Ah! I've been trying to minimize the runtime special-casing for this feature and keep as much of it internalized to the program as possible.  Do you think doing all of this in runtime would be cleaner? Or...less dirty?

Reply:
Cleaner.   There needs to be runtime support regardless

Reply:
Hmm.  OK I'll have to rethink things a bit. It should be down to two sysvars, a TX flag and an instruction inclusion check outside the program with my current thinking.

Off loading execution will decouple the instructions from gas to some extent, right?  Any concerns around that?

Reply:
I looked into moving into runtime and it still seems more straightforward to me for most of this logic to be implemented in the program.

We can do without the second sysvar (which is the part of runtime changes I was most concerned about) if the TX `Message` can be passed one level deeper into `process_instruction`.  This leaves only a test for the `Nonce` instruction in the case that the TX's `recent_blockhash` fails `check_hash_age`.

Reply:
the runtime will do what, exactly, to verify that the TX blockhash exists somewhere?

Reply:
1) `check_hash_age` as per usual
2) if **1** fails, ensure that a `Nonce` instruction exists in the TX.

Reply:
`2.` as described is insufficient, and I think it leads to multiple debits of the fee payer

the runtime needs to fail this TX early and cheaply, there might be something I'm missing here from another conversation, but I think the runtime needs to peek into the Nonce account data before deciding to charge the TX fees.

Reply:
💩 I was afraid it would come to that. So the `tx.msg.recent_blockhash == account.stored_nonce` check will need to come out of program, leaving only the nonce advance logic in the instruction, right?

Lemme dig up which benches this is going to hit before I get ahead of myself...